idx/metric_definition.go

```diff
@@ -0,0 +1,352 @@
+package idx
+
+import (
+	"bytes"
+	"crypto/md5"
+	"fmt"
+	"reflect"
+	"sort"
+	"strings"
+	"unsafe"
+
+	"github.com/raintank/schema"
+	log "github.com/sirupsen/logrus"
+)
+
+//go:generate msgp
+
+type mType uint8
+
+// MType represented by a uint8
+const (
+	MTypeUndefined mType = iota
+	MTypeGauge
+	MTypeRate
+	MTypeCount
+	MTypeCounter
+	MTypeTimestamp
+)
+
+// MetricName stores the name as a []uintptr to strings interned in an object store.
+// Each word is stored as a separate interned string without the '.'. The whole name
+// can be retrieved by calling MetricName.String()
+type MetricName struct {
+	nodes []uintptr
+}
+
+// Nodes returns the []uintptr of interned string addresses
+// for the MetricName
+func (mn *MetricName) Nodes() []uintptr {
+	return mn.nodes
+}
+
+// String returns the full MetricName as a string
+// using data interned in the object store
+func (mn *MetricName) String() string {
+	if len(mn.nodes) == 0 {
+		return ""
+	}
+
+	bld := strings.Builder{}
+	return mn.string(&bld)
+}
+
+func (mn *MetricName) string(bld *strings.Builder) string {
+	// get []int of the lengths of all of the mn.Nodes
+	lns, ok := IdxIntern.Len(mn.nodes)
+	if !ok {
+		internError.Inc()
+		log.Error("idx: Failed to retrieve length of strings from interning library")
+		return ""
+	}
+
+	// should be faster than calling IdxIntern.SetString in a tight loop
+	var tmpSz string
+	szHeader := (*reflect.StringHeader)(unsafe.Pointer(&tmpSz))
```

Comment:
would it make sense for the interning layer to have some `JoinStrings([]uintptr, string) string` method which is basically doing exactly what `strings.Join()` does, but it just takes a `[]uintptr` instead of a `[]string`

Reply:
Yeah something along those lines would be good

Reply:
I agree with that. Seems almost all of the logic here belongs in the interning library.